### PR TITLE
docs, probes: update the probe documentation

### PR DESCRIPTION
**Pull request by jcanocan**
**What this PR does / why we need it**:
Allow to perform the readiness probes. As it was, the SELinux system denied attempts to open the /tmp/healthy.txt file causing the probe to fail. The documentation is modified to explain why the context assignation is required. Finally, the disk image is updated to use a more modern one intended for end users.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Review by xpivarc — COMMENTED** [review]
Looks good I have only one suggestion.

File: docs/probes.md [hunk 1]
```diff
@@ -26,87 +26,78 @@ Make sure to provide enough delay and failureThreshold for the VM and the agent
...
               chpasswd:
                 expire: false
+              password: password
+              user: fedora
               packages:
                 qemu-guest-agent
-              runcmd:
-                - [ "systemctl", "enable", "--now", "qemu-guest-agent" ]
-                - [ "touch" "/tmp/ready.txt" ]
+              runcmd: ['touch /tmp/healthy.txt', 'sudo chcon -t virt_qemu_ga_exec_t /tmp/healthy.txt', 'sudo systemctl enable --now qemu-guest-agent']
```
[the remark this answers is not in the document]
Comment (jcanocan) [reply docs/probes.md:99]: I can rewrite to use one line per command. I agree it's a bit ugly as it is.

File: docs/probes.md [hunk 2]
```diff
@@ -26,87 +26,78 @@ Make sure to provide enough delay and failureThreshold for the VM and the agent
...
-            - masquerade: {}
-              model: virtio
-              name: nic-0
-          networkInterfaceMultiqueue: true
+            - name: default
+              masquerade: {}
           rng: {}
         machine:
-          type: pc-q35-rhel8.2.0
+          type: pc-q35-rhel8.6.0
```
[the remark this answers is not in the document]
Comment (jcanocan) [reply docs/probes.md:70]: Perfect!

File: docs/probes.md [hunk 3]
```diff
@@ -26,87 +26,81 @@ Make sure to provide enough delay and failureThreshold for the VM and the agent
...
+        # Note that timeoutSeconds value does not have any impact before K8s v1.20.
+        timeoutSeconds: 5
       terminationGracePeriodSeconds: 180
       volumes:
+        - containerDisk:
+            image: quay.io/containerdisks/fedora:36
+          name: containerdisk
         - cloudInitNoCloud:
-            userData: |
+            userData: |-
```
Comment (akrejcir) [inline docs/probes.md:91]: Is this change needed? It will remove the newline at the end of the string, but this string is a config file, and usually files end with an empty line.

**Review by akrejcir — COMMENTED** [review]
Thanks for the documentation update.
I have a few minor comments.

Can you also squash the two commits together? I think it would be clearer, because the second commit changes only things modified by the first commit.

File: docs/probes.md (hunk 3 above)
Comment (jcanocan) [reply docs/probes.md:91]: Yes, I are right. Reverted.

File: docs/probes.md [hunk 4]
```diff
@@ -128,3 +122,15 @@ kubectl virt start probe-test
...
+
+```sh
+kubectl get vms -w
+```
+
+6. (optional) Log in to the VM and watch the incoming qemu-ga commands
+
+```sh
+virtctl console readiness-probe
+journalctl -f
```
[the remark this answers is not in the document]
Comment (jcanocan) [reply docs/probes.md]: Following the same thinking as `--type` and `--follow`, I have also changed the `kubectl apply --f /tmp/probe-test.yaml` line by `kubectl apply --filename /tmp/probe-test.yaml`. It may be less necessary than the ones mentioned. Nevertheless, I think it could be necessary for the sake of consistency. Do you agree?

File: docs/probes.md [hunk 5]
```diff
@@ -26,105 +26,111 @@ Make sure to provide enough delay and failureThreshold for the VM and the agent
...
                 expire: false
+              password: password
+              user: fedora
               packages:
                 qemu-guest-agent
               runcmd:
-                - [ "systemctl", "enable", "--now", "qemu-guest-agent" ]
-                - [ "touch" "/tmp/ready.txt" ]
+                - ["touch", "/tmp/healthy.txt"]
+                - ["sudo", "chcon", "--type", "virt_qemu_ga_exec_t", "/tmp/healthy.txt"]
```
Comment (xpivarc) [inline docs/probes.md:101]: Maybe we should leave a note that this will not work without SELinux? Examples are often used in many ways...

Comment (jcanocan) [reply docs/probes.md:101]: Do you mean the `chcon` command or the whole example? If you mean that the `chcon` command will not work, I agree to leave a note telling that they could remove the `chcon` command line or it will fail.

Comment (xpivarc) [reply docs/probes.md:101]: Just the command. Thanks!

Comment (jcanocan) [reply docs/probes.md:104]: Done. I hope the explanation is good enough.

**Review by xpivarc — COMMENTED** [review]
/approve

**Comment by dominikholler** [comment]
@akrejcir Can you please ack that all your comments are addressed?
@xpivarc What is missing to get this PR merged? Has the CI to be triggered manually, because @jcanocan is not yet part of the kubevirt org?

**Comment by xpivarc** [comment]
> @akrejcir Can you please ack that all your comments are addressed? @xpivarc What is missing to get this PR merged? Has the CI to be triggered manually, because @jcanocan is not yet part of the kubevirt org?

A LGTM needs to be applied and then it is just about tests.

File: docs/probes.md (hunk 3 above)
Comment (akrejcir) [reply docs/probes.md:94]: You probably missed this.

File: docs/probes.md [hunk 6]
```diff
@@ -26,105 +26,114 @@ Make sure to provide enough delay and failureThreshold for the VM and the agent
...
           interfaces:
-            - masquerade: {}
-              model: virtio
-              name: nic-0
-          networkInterfaceMultiqueue: true
+            - name: default
+              masquerade: {}
           rng: {}
         machine:
           type: pc-q35-rhel8.2.0
```
Comment (lyarwood) [inline docs/probes.md:73]: I've got no idea why this is in an upstream example as it's only shipped downstream in RHEL.

File: docs/probes.md [hunk 7]
```diff
@@ -26,105 +26,114 @@ Make sure to provide enough delay and failureThreshold for the VM and the agent
...
                 expire: false
+              password: password
+              user: fedora
               packages:
                 qemu-guest-agent
               runcmd:
-                - [ "systemctl", "enable", "--now", "qemu-guest-agent" ]
-                - [ "touch" "/tmp/ready.txt" ]
+                - ["touch", "/tmp/healthy.txt"]
+                - ["sudo", "chcon", "--type", "virt_qemu_ga_exec_t", "/tmp/healthy.txt"]
```
Comment (lyarwood) [inline docs/probes.md:104]: I get listing this as a caveat but if it's already installed in the referenced Fedora image then why add the commands here?

File: docs/probes.md [hunk 8]
```diff
@@ -26,105 +26,114 @@ Make sure to provide enough delay and failureThreshold for the VM and the agent
...
+            - name: containerdisk
+              disk:
                 bus: virtio
-              name: cloudinitdisk
-            - bootOrder: 1
+            - name: cloudinitdisk
               disk:
                 bus: virtio
-              name: rootdisk
           interfaces:
```
Comment (lyarwood) [inline docs/probes.md:68]: You can drop `interfaces` and `networks` if all you want is an interface on the default network. One will be added automatically.

File: docs/probes.md [hunk 9]
```diff
@@ -26,105 +26,114 @@ Make sure to provide enough delay and failureThreshold for the VM and the agent
...
           name: cloudinitdisk
-        - containerDisk:
-            image: kubevirt/fedora-cloud-container-disk-demo
-          name: rootdisk
 ```
 2. Apply the VM manifest
 
 ```sh
-kubectl apply -f /tmp/probe-test.yaml
+kubectl apply --filename /tmp/probe-test.yaml
```
Comment (lyarwood) [inline docs/probes.md:111]: I'm not sure if this is worth changing tbh, if anything you can just merge steps 1 and 2 into a single apply step.

File: docs/probes.md [hunk 10]
```diff
@@ -26,105 +26,114 @@ Make sure to provide enough delay and failureThreshold for the VM and the agent
...
 ```
 
 3. Start the VM
 
 ```sh
-kubectl virt start probe-test
+virtctl start readiness-probe
```
Comment (lyarwood) [inline docs/probes.md:117]: As above you can just mark the VM to always run in the definition and merge this into step 1.

File: docs/probes.md [hunk 11]
```diff
@@ -26,105 +26,114 @@ Make sure to provide enough delay and failureThreshold for the VM and the agent
...
 # This will stream the events including any probe failures.
-# Observe the guest-agent becomming available here.
+# Observe the guest-agent becoming available here.
 kubectl get events --watch
 ```
 
+5. Check if the `.status.ready` field is `true`, it may take a bit
+
+```sh
+kubectl get virtualmachine readiness-probe --watch --output jsonpath='{.status.ready}{"\n"}'
```
Comment (lyarwood) [inline docs/probes.md:131]: Could `kubectl wait vms/readiness-probe --for=condition=Ready` work here?

**Review by lyarwood — CHANGES_REQUESTED** [review]
Apologies I should've held off on marking this ok-to-test, assumed it was ready to merge.

There are a few changes required here, the use of a downstream machinetype being the most concerning tbh.

**Comment by lyarwood** [comment]
Unrelated to this review but could you also give the examples in the user-guide a once over if you have time?

https://github.com/kubevirt/user-guide/blob/main/docs/virtual_machines/liveness_and_readiness_probes.md

**Comment by jcanocan** [comment]
> Unrelated to this review but could you also give the examples in the user-guide a once over if you have time?
> 
> https://github.com/kubevirt/user-guide/blob/main/docs/virtual_machines/liveness_and_readiness_probes.md

Yes, I'm currently working on that :)

File: docs/probes.md (hunk 9 above)
Comment (jcanocan) [reply docs/probes.md:111]: In my humble opinion, I believe it is more educational to have all steps. Please, feel free to disagree, it's just my opinion.

File: docs/probes.md (hunk 10 above)
Comment (jcanocan) [reply docs/probes.md:117]: Same as above.

File: docs/probes.md (hunk 9 above)
Comment (lyarwood) [reply docs/probes.md:111]: Yeah understood and not something I'd block on. IMHO however making examples as minimal as possible helps readers focus on the relevant content of the example, in this case the `readinessProbe` and not the lifecycle of a VM.

File: docs/probes.md (hunk 11 above)
Comment (jcanocan) [reply docs/probes.md:131]: Yes, it works. However, in my humble opinion, I think it's better for learners to spectate the .status.ready changes than just wait for it to meet the condition. What do you think?

File: docs/probes.md (hunk 9 above)
Comment (jcanocan) [reply docs/probes.md:111]: Ok, I agree. I will merge steps 1 and 2. Thanks for the inputs!

File: docs/probes.md (hunk 6 above)
Comment (jcanocan) [reply docs/probes.md:73]: I'm not sure either tbh. @xpivarc, could you give us some input on this? Thanks!

File: docs/probes.md (hunk 7 above)
Comment (jcanocan) [reply docs/probes.md:104]: I'm not sure if I fully understand your question. But without the `chcon` command, the probe will fail because SELinux will deny attempts to open the `/tmp/healthy.txt` because of the context. Sorry, maybe this is not what you are asking.

File: docs/probes.md (hunk 3 above)
Comment (jcanocan) [reply docs/probes.md]: Apologies, I missed that. Now it's reverted to the original state.

File: docs/probes.md (hunk 6 above)
Comment (xpivarc) [reply docs/probes.md:70]: I am not sure but  I meant we should remove this here `https://github.com/kubevirt/kubevirt/pull/8400#discussion_r960874142`

Comment (jcanocan) [reply docs/probes.md]: Thanks for the input! Removed.

File: docs/probes.md (hunk 11 above)
Comment (lyarwood) [reply docs/probes.md:112]: I understand what you're saying but that isn't common practice, using `wait` and `--for` is. I'll not block on this but I'd still rather use them over this command.

File: docs/probes.md (hunk 7 above)
Comment (lyarwood) [reply docs/probes.md:96]: I think I've misunderstood your following comment at the top of the file:

> Also, cloud-init assigns the proper SELinux context, i.e. virt_qemu_ga_exec_t, to the `/tmp/healthy.txt` file. 

I assumed you were saying `cloud-init` was doing this for us automatically somehow.

**Review by lyarwood — COMMENTED** [review]
/lgtm

File: docs/probes.md (hunk 7 above)
Comment (jcanocan) [reply docs/probes.md:96]: Maybe I can extend a bit this explanation to make it cleaner.

File: docs/probes.md (hunk 11 above)
Comment (jcanocan) [reply docs/probes.md:112]: I see. I think you are right, I've seen more often `--for`, so I will change it. Thanks!
